metals/src/main/scala/scala/meta/internal/metals/ScalafmtConfig.scala

```diff
@@ -0,0 +1,148 @@
+package scala.meta.internal.metals
+
+import java.nio.file.FileSystems
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+import scala.meta.io.AbsolutePath
+
+import com.typesafe.config.Config
+import com.typesafe.config.ConfigFactory
+import com.typesafe.config.ConfigParseOptions
+import com.typesafe.config.ConfigSyntax
+import com.typesafe.config.parser.ConfigDocument
+import com.typesafe.config.parser.ConfigDocumentFactory
+
+object ScalafmtConfig {
+
+  sealed trait SyntaxError
+  case object SyntaxError extends SyntaxError
+
+  case class FileOverride(
+      glob: String,
+      dialect: ScalafmtDialect
+  )
+
+  case class ConfigValues(
+      version: Option[String],
+      runnerDialect: Option[ScalafmtDialect],
+      fileOverrides: Map[String, ScalafmtDialect]
+  ) {
+
+    def overrideFor(path: AbsolutePath): Option[ScalafmtDialect] = {
+      val fs = FileSystems.getDefault
+      fileOverrides
+        .find { case (pm, _) =>
+          val matcher = fs.getPathMatcher(pm)
+          matcher.matches(path.toNIO)
+        }
+        .map(_._2)
+    }
+  }
+
+  def update(
+      configText: String,
+      version: Option[String],
+      runnerDialect: Option[ScalafmtDialect]
+  ): String = {
+
+    def docFrom(s: String): ConfigDocument = {
+      val options = ConfigParseOptions.defaults().setSyntax(ConfigSyntax.CONF)
+      ConfigDocumentFactory.parseString(s, options)
+    }
+
+    def withUpdatedVersion(content: String, v: String): String = {
+      val doc = docFrom(content)
+      if (doc.hasPath("version"))
+        doc.withValueText("version", '"' + v + '"').render
+      else {
+        // prepend to the beggining of file
+        val sb = new StringBuilder
+        sb.append(s"version = $v")
+        sb.append(System.lineSeparator)
+        sb.append(content)
+        sb.toString
+      }
+    }
+
+    def withUpdatedDialect(content: String, d: ScalafmtDialect): String = {
+      val doc = docFrom(content)
+      if (doc.hasPath("runner.dialect"))
+        doc.withValueText("runner.dialect", d.value).render
+      else {
+        // append to the end
+        val sb = new StringBuilder
+        sb.append(content)
+        val sep = System.lineSeparator
+        val lastLn = content.endsWith(sep)
+        if (!lastLn) sb.append(sep)
+        sb.append(s"runner.dialect = ${d.value}")
+        sb.append(sep)
+        sb.toString
+      }
+    }
+
+    val applyVersion =
+      version.fold(configText)(withUpdatedVersion(configText, _))
+    runnerDialect.fold(applyVersion)(withUpdatedDialect(applyVersion, _))
+  }
+
+  def parse(path: AbsolutePath): Either[SyntaxError, ConfigValues] =
+    parse(ConfigFactory.parseFile(path.toFile))
+
+  def parse(text: String): Either[SyntaxError, ConfigValues] =
+    parse(ConfigFactory.parseString(text))
+
+  def parse(config: Config): Either[SyntaxError, ConfigValues] = {
```

Comment:
Just a question. Does Scalafmt expose this functionality that we could just utilize instead of adding it all in here? This is a lot of logic to parse the scalafmt file, and then all that logic will probably be re-done in Scalafmt again later on. It'd be great to just utilize that.

Reply:
Scalafmt doesn't normally modify the file, so that will not be available in the API. 

Alternatively,  we could just append file override to the file as you can have multiple and they should be merged in HOCON (I think). As for version and dialect, we could use regexes. 

However, if we already have this logic here then maybe I would leave it.

Reply:
For parsing config, we would need to depend on Scalafmt and metaconfig artifacts themselves, so I think it's better to avoid it.